### DISCUSSION TITLE
How to handle grain merging for grains with same crystal orientation in `GrainTracker` ？

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#22853] [Q&A Modules: Phase field] [answered]
Dear MOOSE experts,

Recently, I wanted to perform grain merging based on `GrainTracker`, when two grains have the same crystallographic orientation, or their misorientation is below a very small value. But I'm not sure how such a merge operation should be performed in GrainTracker.

Specifically, when I was performing my EBSD microstructural grain growth simulation, I found that there was a twin that split the matrix grain into two grains, as shown in the image below.
![image](https://user-images.githubusercontent.com/56183775/205245168-398d4085-a105-4c3e-b212-43f01cfbce9e.png)

When I use the novel feature extraction and remapping algorithm developed by @permcody, due to the order parameter remapping operation, even if the twin disappears, there are still two grains left. Instead of merging into a complete grain under the same orientation as in the experiment. So I want to perform the operation of merging grains in `GrainTracker` based on the direct orientation relationship of the two grains.

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by GiudGiud** [comment]
Hi
There s documentation on the grain tracker on this
https://mooseframework.inl.gov/moose/source/postprocessors/GrainTracker.html
which part in particular would you want more information on?

Are you planning to modify the GrainTracker?

Guillaume

**Reply by PengWei97** [reply]
Oh, yeah. Thank you very much for your reply. I have read all the public information about GrainTracker, including web, cody's two papers and graduation thesis. Of course, please forgive me for still having a lot of confusion here.

> Are you planning to modify the GrainTracker?

Yes, in order to accomplish the purpose mentioned above, I plan to further build a class based on `GrainTracker` to determine whether the grain needs to be merged according to the orientation relationship of adjacent grains (from the input EBSD data).

Wei

**Comment by permcody** [comment, marked as answer]
Hi Wei,

Really happy to see this capability being exercised. It's not something I really had to work with much when I was working on this problem, but it did pop up in a few of our data sets. Briefly, there are two edge cases that have to be dealt with when working with "twins" (grains with the same crystallographic orientation). There's the merge case when they come together during evolution (is this even physical?) and the case where they are created due to the right combination of an odd grain shape and the right kinds of physics (again, I can't comment too much on how physical this is or not). I attempted to handle both in the snippet linked below. The way this is handled is by making sure the grains really are a twins (see the criteria starting on line 721). Ultimately, they should be represented by the same grain ID even though you have two (or more) physically disconnected features. Once two features are marked with the same ID, the GrainTracker should not attempt to remap either one of them if they get close enough to coalesce during a simulation. That's really all there is too of it from a logical point of view. I'm not sure I understand the rest of your inquiry about "one disappearing" but having "two left". It's true that the grain tracker never removes an inactive grain from the data structure, but anything that shrinks or joins together does get marked as "INACTIVE" and ignored by the GrainTracker for the remainder of the simulation. 

https://github.com/idaholab/moose/blob/6099fe533a9424e9fd3a73ecac6cfeb2eda6eb89/modules/phase_field/src/postprocessors/GrainTracker.C#L663-L738

There is also a test you can look at here:

https://github.com/idaholab/moose/blob/next/modules/phase_field/test/tests/grain_tracker_test/split_grain.i

I'm not sure if I've answered your question, but take a look at the logic and the test and see if that's helpful. It's possible that this may not work exactly as you need it to, but we can discuss possible modifications to make it work for your needs.

Cody

**Reply by PengWei97** [reply]
Hi Cody,

I really appreciate your kind and detailed advice. Today I carefully read your suggestion, but I has some confusion about your reply, and I hope to get your further advice.

> There's the merge case when they come together during evolution (is this even physical?)

We can find from the above figure that the merging of two substrate  grains is physical, when the intermediate twin disappears. I would like to ask whether ` they come together` here refers to the merger of twins or the merger of two substrate grains?


> the case where they are created due to the right combination of an odd grain shape and the right kinds of physics (again, I can't comment too much on how physical this is or not).

Also what does `they` refer to here? And my current processing method is to directly import the initial EBSD structure into moose and use GrainTracker to track grains, in which the shape and type of twins are determined based on Euler angles.


> I'm not sure I understand the rest of your inquiry about "one disappearing" but having "two left".

Yes, your understanding is correct.  
Here's what I want to do based on your suggestion to accomplish grain merging，considering the difficulty of accurately identifying twins, I think the calculation of misorientation (based on the Euler angles of EBSD) can be performed for each grain and its adjacent grains. When their misorientation is lower than a very low value, the operation of `grain._id = other_grain._id` is executed to make the two grains merge. (Such merging also has physical meaning, because when the orientation difference between two grains is 0, they should be the same grain).

The above is my confusion about your enthusiastic suggestion and the operation I want to perform with reference to the code you gave, and I look forward to your next suggestion.

Wei

**Reply by PengWei97** [reply]
Again, thank you very much for recommending such an important code block to me, considering that I need to add some new code on top of GrainTracker, so I feel that it is necessary to understand this code completely correctly, the following is my understanding of this code Understood, hope to get your affirmative answer or correction of misunderstanding,

For Case 1, traverse `_feature_sets` in the new set, and perform the following operations when `grain._status == Status::CLEAR`
1. If it is determined as `Splitting grain`, then set `grain._id = other_grain._id`. Among them, `grain` is the first part split from a **mother grain** in the old set, called **grain A**, and it does not match the new set; the other part of the split is called `other_grain` or **grain B**, and its state in the new set is set to `Status::MARKED`,  **grain B** is the same as the parent grain in the old set. Referring to the picture in your article, according to my understanding, I made the following annotations，
![image](https://user-images.githubusercontent.com/56183775/205443491-15e5ff18-8831-4faf-8d7a-0035ef066a13.png)

2. If (grain._var_index < _reserve_op_index),
    1. If only one `old_grain` in `_feature_sets_old` matches the current grain, then `grain._id = _feature_sets_old[old_grain_indices[0]]._id;`
    2. If there are more than one, `Need more information to find correct candidate - contact a developer!`
4. If the above conditions are met, then `grain` must be a nucleating grain.

For Case 2, I don’t think there is much to explain. Its purpose is to make the state of `Status::CLEAR` in `_feature_sets_old` change to `Status::INACTIVE` in `_feature_sets_old`.

The above is my understanding of this code, looking forward to your guidance.

Happy weekend.

Wei

**Reply by permcody** [reply]
Hi Wei,

1. This is mostly correct. However, rather than discarding the "mother grain" and replacing it with two new grains ("A" and "B"), what happens with the algorithm as written is that both "Grain A" and "Grain B" will attempt to "match up" with the "Mother Grain" from the previous step. This is because at the point where they physically separate in a time step, they'll both have the same order parameter at the start of that time step, as as the middle picture of the figure indicates, they "should" have halos that touch. In that case, we'll allow one of the two grains to match up to the mother grain (doesn't really matter which one), and then the remaining grain will be "detected as a split grain" due to the criteria that I just explained. What's important though is that this isn't a case of merging twins, this is the opposite, this is a potential mechanism for creating twins (splitting)!

2. 
    i. This is exactly the case I just mentioned, the slow splitting of a grain into two child grains.  
    ii. This is a case where we have more than two child grains (with overlapping halos) in a single time step. This would be something like a 3-way split. I didn't think this was likely to happen so I just punted and decided to print out a message to contact the developer. I don't think you will hit this case.

3. Nucleation - you aren't dealing with that at all in your current sim. All of that logic only kicks in if you start using the "reserve OP" variables.

So back to your original question. Who to deal with twins merging?

I was just reviewing the code again and realized that I specifically setup the GrainTracker to NOT treat twins in any special way (e.g. I made sure to remap one of them so that if they "got too close", one of them would be remapped to a new order parameter preventing them from re-merging. So you are correct, you (we) will need to adjust the grain tracker if you do indeed want to handle this case. So first order of business. I don't believe any of the logic in the "trackGrains" routine will need to change. Once we get the code to a place where twins re-merge, it'll be equivalent to a grain disappearing as it shrinks. Essentially in one time step, we'll have two child grains that move together to form one new grain, triggering the logic to handle the "inactive" grain case. Instead what needs to specifically change is how we handle twins in the remapping phase. Take a look at the logic here:

https://github.com/idaholab/moose/blob/af38e9f0b6976df8d0a90b1c6b339bef1dbc9b00/modules/phase_field/src/postprocessors/GrainTracker.C#L956-L976

In this logic, we are specifically dealing with twins (we know this because they share an id), but we are still remapping one of them to a new order parameter. I think the easiest thing to do would be to add a new Boolean to the GrainTracker (GrainTrackerInterface) that would prevent twins from getting remapped. Something like`remap_twins = false`, then in the lines above, we simply don't remap one twin if the grain that's intersecting is the other twin.

The tricky bits: Implementing and testing out that new logic should be pretty easy if you have the right data set. The problem is that once you allow twins to not remap you open up several new edge cases. The first being that if you end up needing to remap a twin due to being close to an unrelated grain sharing the same order parameter, you'll need to remap the twin too and make sure there aren't any conflicts. You'll also have to add new logic to the "trackGrains" routine to essentially ignore messages that you may otherwise encounter from twins being close in subsequent time steps. 

My suggestion is that you start with the Boolean and ignore the remap and see what happens. Even with the new edge cases this should be mostly working just with the new Boolean!

**Reply by permcody** [reply]
Ok... Rereading that code block again for the third time... I'm still not getting this right. Look at the console message on line 963. It's explicitly trying to make sure the twins are on the same order parameter. This will allow them to merge if they get close enough. So, I'm again thinking that the merging case you are looking at really should be working already. I think I'll need to see a counter-example to provide better feedback.

**Reply by PengWei97** [reply]
Hi Cody,

> ... this is a potential mechanism for creating twins (splitting)!

1. I totally agree with you. Case1 (for Splitting Grain) is the potential function with splitting twins.


> Once we get the code to a place where twins re-merge, it'll be equivalent to a grain disappearing as it shrinks.

Yes, when it is determined that two grains (such as grain_i and grain_j) need to be merged, it is essentially to merge grain_j into grain_j (as you said one grain disappears), even if the order parameter indexes of the two grains are different.


> Essentially in one time step, we'll have two child grains that move together to form one new grain, triggering the logic to handle the "inactive" grain case.

Yeah, I think it is important to note that we are performing a merge operation and the new big grain has the same grain ID and `_var_index` as some child grain.


> The problem is that once you allow twins to not remap you open up several new edge cases. 

Yes, there may be some unexpected situations, such as merging grains to the same order parameter, this new large grain may have the same order parameter as the surrounding grains, this is what I think of the possible situation.


******

**Ok, I tried to focus the Key problem on solving the merging of two grains, even though they have different order parameters.** Specifically, when two grains (e.g. grain_i and grain_j) are identified and their order parameter indexes (e.g. gr1 and gr5) are different, how to realize the re-merger operation of these two grains? 

Next, based on your suggestion, the `mergeGrainsMisorientation()` function was created to do this in `GrainTracker::trackGrains()`, but it didn't meet our expectations, that is, `Creating new unique grain` in the next timestep. The core code in `mergeGrainsMisorientation()` is as follows,

```c++
void 
GrainTracker::mergeGrainsMisorientation()
{
  // The simplest example: focus on specifying the merging of two grains to test how to re-merge grains
  // For example, specifying to merge grain_i (_id == 0) and grain_j (_id == 2), and they have different order parameter indexs. Then, set `grain_j._id = grain_i._id` and `grain_j._id = grain_i._id`
  for (const auto i : index_range(_feature_sets))
  {
    for (const auto j : index_range(_feature_sets))
      if (_feature_sets[i]._id == 0 && _feature_sets[j]._id == 2)
      {
        // only make the Grain id of grain_j become the Grain id of grain_i,
        // And the order parameter index modification operation is executed in remap algorithm;
        std::cout << "merging grain_i " << _feature_sets[i]._id << " and grain_j " << _feature_sets[j]._id << std::endl;     
        std::cout << "merging grain_i._var_index " << _feature_sets[i]._var_index << " and grain_j._var_index " << _feature_sets[j]._var_index << std::endl;   

        _feature_sets[j]._id =  _feature_sets[i]._id;
        _feature_sets[j]._var_index =  _feature_sets[i]._var_index;
        std::cout << "the grain_j._id " << _feature_sets[j]._id << std::endl;
        std::cout << "the grain_j._var_index " << _feature_sets[j]._var_index << std::endl;
        break;

        // grain_adj._status = Status::INACTIVE;     
      }
  }  
}
```

And the terminal outputs the following information, indicating that the merging of grains is not successful,
```
...
Time Step 0, time = 0
...

Time Step 1, time = 0
...
merging grain_i 0 and grain_j 2
merging grain_i._var_index 1 and grain_j._var_index 5
the grain_j._id 0
the grain_j._var_index 1
Finished inside of GrainTracker
...

Time Step 2, time = 10.5, dt = 5.5
...
[32mMarking Grain 0 as INACTIVE (variable index: 1)
[39m[33mNucleating Grain Detected  (variable index: 5)
[39m[33m
*****************************************************************************
Couldn't find a matching grain while working on variable index: 5
Creating new unique grain: 10
Grain ID: 10
...

Finished Executing    
```
So, in general, I have tried to achieve the simplest grain merging function these days, without considering the remap algorithm. However, even though the merge operation was performed at `Time Step 1`, it was recognized as a new grain at `Time Step 2`, or because `grain_j._var_index` has not changed.  In addition, I also provide the complete code modification for [GrainTrack ](https://github.com/PengWei97/moose/commit/950550336aad0c1d46dfdca39f35084f1480b181#diff-30e06aed62e291614ee5c60953e7f4b4008ec7a1cbbc8379abfa0479a080d12b)and input file ([vt_10.i](https://github.com/PengWei97/luwu/commit/07ed5d282fa728a4894aa3cdf202c84351de525a#diff-911f733b13f84144d879972dcfe028b3abf26d2e65b6ffdaa598a207a83ea456)) to prevent me from missing key information.

**Finally, my question is how to modify the operation of merging grains.** And it seems that even if `_feature_sets[j]._var_index =  _feature_sets[i]._var_index;` is executed, `_feature_sets[j]._var_index` is still 5 at `Time Step 2`;

Thank you very much for your enthusiastic guidance and look forward to your next feedback.

Wei

**Reply by permcody** [reply]
Wei,

Sorry it's taken me so long to get back to you. I still need to think about this a bit more, but I want to at least communicate the purpose and intent of a few of these function to suggest what you can look at to resolve this behavior.

You mention here at the bottom of your thread that the twin is recognized as a new grain in time step 2. That piece of information is critical because once the GrainTracker "thinks" it's a separate grain, it'll certainly make sure that it stays apart from every other grain so things will not re-merge as you've tried here. You can't force the re-merge outcome in the GrainTracker at all, it acts more like a Postprocessor than a "Kernel". Instead, the key to getting this to work properly is to setup the right conditions so that remerging will occur naturally during the solution phase. Essentially this will happen automatically if any two grains with the same order parameter get "too close". 

Again, without looking at the code, what needs to happen is that information about twins is preserved so that we can force them to be on the same order parameter all the time. This means that any time any twin gets remapped to a different order parameter, we'll want to make the twin follow as well. In my opinion, this is all that needs to happen.

I need to find some time to look at the code as I was pretty sure that I had already done this. It's possible the algorithm just has a bug, or perhaps I just didn't finish? I thought about these cases back when I implemented these algorithms, but I only had a single test case that I created. The same ones that I used to generate those images for the splitting and remerging cases. 

I'll let you know if I see anything that should be investigated further.
Best Regards,
Cody

**Reply by PengWei97** [reply]
Hi permcody,

I believe that the grain merging problem has now been successfully resolved, with some modifications made to GrainTrack. Below is my approach and code for addressing the issue, although it may not be perfect. I would appreciate any feedback or comments you may have.

> You can't force the re-merge outcome in the GrainTracker at all, it acts more like a Postprocessor than a "Kernel". Instead, the key to getting this to work properly is to setup the right conditions so that remerging will occur naturally during the solution phase.

Yes, I customized the `GrainTrackerGG2` to handle grain merging operations only when certain conditions are met, such as when the misorientation angle is less than 1°.

Below are the three steps or stages involved in accomplishing the merging process:
1. Initially, I created GrainTrackerGG by building upon GrainTracker. Then, I developed the `createAdjacentIDVector()` function to [obtain the vector of each grain and its adjacent grain ID](https://github.com/PengWei97/yinglong/commit/8d14d53607e842cf621b2386ad11f7a731cb1c79);
2. Afterwards, I created GrainTrackerGG2, which inherits from GrainTrackerGG, and I[ overrode the trackGrains and remapGrains methods](https://github.com/PengWei97/yinglong/commit/f9eff63242fc1ffc7a973a0563df6839df296892).
3. Finally, I added [the merging functionality](https://github.com/PengWei97/yinglong/commit/3bbbf38ab926cf58ea1539180a26b44f84c80345) to GrainTrackerGG2.

The following code is primarily responsible for handling the merging operation in [GrainTrackerGG2](https://github.com/PengWei97/yinglong/blob/93636a9b268724162000e428c55b5192c993414b/src/postprocessors/GrainTrackerGG2.C):
```c++

void
GrainTrackerGG2::remapGrains()
{
  ...
     // Case 2 (inactive grains in _feature_sets_old)
    for (auto & grain : _feature_sets_old)
    {
       ...
            /**
             * We're not going to try very hard to look for a suitable remapping. Just set it to
             * what we want and hope it all works out. Make the GrainTrackerGG great again!
             */
            grain1._var_index = grain2._var_index;
            grain1._status |= Status::DIRTY;

            if (_remerge_grains)  // by weipeng
              grain_id_to_new_var.emplace_hint(
                  grain_id_to_new_var.end(),
                  std::pair<unsigned int, std::size_t>(grain1._id, grain1._var_index));
       ...
     }
  ...
}

void 
GrainTrackerGG2::mergeGrainsBasedMisorientation()
{
  ...
      if (misor_angle < threshold_merge)
      {
        _console << COLOR_YELLOW << "Grain #" << grain_i._id << " and Grain #" << grain_j._id
                 << " was merged (misor: " << misor_angle << ").\n"
                 << COLOR_DEFAULT;

        grain_j._id = grain_i._id;
      }
  ...
}
```
The result of the execution is as follows,
![image](https://user-images.githubusercontent.com/56183775/225201599-395a8cf9-d127-4fc0-9040-152f4a08bbba.png)

Finally, I would like to express my sincere gratitude for your generous assistance, which helped me to successfully complete this task. I would greatly appreciate any feedback or suggestions you may have regarding this aspect. Additionally, I would be more than happy to merge this functionality into GrainTracker if it would be useful.

Best Regards,
Wei

**Reply by GiudGiud** [reply]
@permcody

**Reply by permcody** [reply]
Wei,

This is fantastic work! I'm very happy to see that you've managed to enhance the capabilities for your need. If you are interested, I would invite you to add your contributions to the MOOSE repository. I believe that the code above is reasonable to add to the existing GrainTracker class. Provided it be activated with appropriate input parameters, which it appears you have already done. I would also entertain adding a new inherited class provided we minimize copying from the existing class, even if that means we need to add additional hooks in the existing class to support your implementation. In either case, we are interested in this work.

I would ask for the following:
- Your new and modified code changes
- One or more regression tests that exercise your new logic
- Documentation: Either added to the existing `GrainTracker.md` markdown, or a new Markdown file if you decide to contribute a new class.

We can assist you throughout the process!
Best Regards,
Cody

**Reply by PengWei97** [reply]
Permcody,

It is a great honor to receive your approval, and I am excited about merging my code into Moose. Thank you for the opportunity. I would also appreciate your guidance on the standard process of code development. As a Moose rookie, I am eager to learn and improve.  

Next, I will provide a detailed explanation of the additions and modifications I made to `FeatureFloodCount` and `GrainTracker`. And I also created PR #23757 to merge this part of the code.

[First](https://github.com/idaholab/moose/pull/23757/commits/7fcc36d04d25a0923c44910378fd9d5cca6c622e), I added `adjacent_id` to the `FeatureData` class in `FeatureFloodCount`. This is used to store the grain ID of each adjacent grain for each feature or grain. It should be noted that the storage of the vector of adjacent grain IDs is done in GrainTracker. Additionally, I created three functions, `getFeatureID, getAdjacentID, and getAdjacentGrainNum`, to extract relevant information. 

But here there are some warnings, and I don't know how to eliminate them,
```bash
/home/pw-moose/projects/moose/modules/phase_field/build/header_symlinks/FeatureFloodCount.h:324:26: warning: 'FeatureFloodCount::FeatureData::_boundary_intersection' will be initialized after [-Wreorder]
  324 |     BoundaryIntersection _boundary_intersection;
      |                          ^~~~~~~~~~~~~~~~~~~~~~
/home/pw-moose/projects/moose/modules/phase_field/build/header_symlinks/FeatureFloodCount.h:305:31: warning:   'std::vector<unsigned int> FeatureFloodCount::FeatureData::_adjacent_id' [-Wreorder]
  305 |     std::vector<unsigned int> _adjacent_id;
      |                               ^~~~~~~~~~~~
/home/pw-moose/projects/moose/modules/phase_field/build/header_symlinks/FeatureFloodCount.h:171:5: warning:   when initialized here [-Wreorder]
  171 |     FeatureData(std::size_t var_index,
```

[Second](https://github.com/idaholab/moose/pull/23757/commits/5a4141ef5a5a99bb5ed5d8f652d8d454e7e62758), add `createAdjacentIDVector()` in `GrainTracker.h` to complete the purpose of creating adjacent grain vector;

[Third](https://github.com/idaholab/moose/pull/23757/commits/cd35df0ca2869e88f5e0533b2ed24d9bf2f75e22),  we need to add an interface for grain merging by creating a virtual function called "mergeGrainsBasedMisorientation" in the GrainTracker. I am planning to create a derived class, GrainTrackerMerge, which will be responsible for performing the detailed operations of grain merging based on the calculation of misorientation.

[Fourth](https://github.com/idaholab/moose/pull/23757/commits/3d1ad7fc6000cb0fa2901af36e5e1b9dd27238d6), add the function of grain merging to the derived class GrainTrackerMerge, based on the calculation of misorientation. But after the creation is complete, I run a test and it shows an error like this,
```bash
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 2945348 RUNNING AT pwmoose-PowerEdge-T640
=   EXIT CODE: 11
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```
I suspect that the error may be related to creating the derived class GrainTrackerMerge based on GrainTracker. Although it may seem like a simple mistake, I would appreciate your guidance on this issue.

[Fifthly](https://github.com/idaholab/moose/pull/23757/commits/357bd0d44d2c53c4296e4bdae1f561b5abce2317), as I was unable to resolve the aforementioned bug, I added the function of grain merging directly to GrainTracker. Although I recognize that this may not be the wisest decision, I intend to move the last part of the code to the derived class GrainTrackerMerge once the bug is resolved；

[Sixthly](https://github.com/idaholab/moose/pull/23757/commits/81fe6e87bce2d69c2a3ba8a530d688a010e83668), I created a test to verify the newly modified GrainTracker, which I named "test_GrainTrackerMerge.i", and placed it in the directory "./modules/phase_field/test/tests/GrainTrackerMerge". However, when I attempted to run the test, I encountered the following error:

```bash
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 21

*** ERROR ***
A 'EulerAngleFileReader' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.
```
Afterwards, I copied the "test_GrainTrackerMerge.i" file to my own APP-yinglong directory and ran the test, which passed successfully. The results of the operation are shown below:
![image](https://user-images.githubusercontent.com/56183775/225634152-9659e99e-b9f7-42bc-b369-9a9d97670c0f.png)


Overall, I believe that I have successfully added the misorientation-based grain merging functionality to the grain tracker algorithm. I made some modifications and additions to the FeatureFloodCount and GrainTracker classes, and created a derived class, GrainTrackerMerge, for specific implementation. However, I hope to get your assistance in resolving the two bugs mentioned above.

Best Regards,
Wei

**Reply by permcody** [reply]
1. I reviewed your PR and mentioned how to fix this. Bottom line, your declaration and initialization must be in the same relative order to avoid this warning.
2. I like this, however I made a comment in your PR that we may not always want to populate that vector. I suggest you only populate it in your new class (or trigger the population of it from your new class). It can be expensive since it's a quadratic algorithm wrapping either a linear or n log n algorithm. We don't usually need it populated.
3. Yes! This is what I'd like to see, just a new hook in the base class with all the new code in your new class
4. Seg faults are the best kinds of errors to track down. Follow the instructions here to look for the bug: https://mooseframework.inl.gov/application_development/debugging.html
5. I see, I suggest returning to your first implementation and see if you can find it with debugging guidance. Seg faults usually happen because you try to "dereference" something that hasn't be initialized. We might just see it from the code itself but if not, the debugger will pinpoint it for you.
6. That particular object is registered in the `TensorMechanics` module, but Tensor Mechanics is compiled into the `PhaseField` module so it should be available. I suggest you push up your test and let it fail. We should be able to look at it more to find the issue.

Hopefully between the review of your PR and the feedback here you have enough to continue on.

**Reply by permcody** [reply]
Looking at your result, why are Grain 1 and Grain 2 separate to begin with? I would think a better test would be to start with an IC closer to the image you showed at the start of this discussion (the two grains separated by a tiny wedge). I'm very interested in such a test as I still think the GrainTracker should "naturally" already handle that case or should be very close to it. When I designed it, I intended to support having different grains in the domain with the same ID. It was my intent that if those grains were to come into contact later in the simulation, they would just join (because the GrainTracker is *not* supposed to remap their underlying order parameters in that case). 

Admittedly, I don't believe we have an integration test for that case so there may be issues with it, but it should work. I recognize that your work is still useful in that ideally you'd want different physics for joining grains that have slight misorientations too. 

The design I'd like to move towards would be to have the GrainTracker just make a callback that could be overridden in a derived class that could ultimately make the decision on whether two grains that "are coming into proximity to one and other to be remapped or not". That decision could be made based on orientation, or simply on initial ID (such as in the case or reading in an EBSD data set, or whatever other physics makes sense). The GrainTracker shouldn't be making that decision on it's own. I want it to be more general. This is why I like you creating a new class to do this work.

Thanks again.

**Reply by PengWei97** [reply]
Yes, you mentioned two scenarios: one is when grains with the same orientation are initially adjacent, but are recognized as two separate grains by EBSD recognition software (such as MTEX, Dream3D); the other is when grains with the same orientation are not initially adjacent. However, it should be noted that the test I provided only shows that when two grains have the same orientation, a merge operation is necessary based on the GrainTrackerMerge class. Furthermore, I believe that the two scenarios you mentioned are essentially the same, as they both involve merging adjacent grains with the same orientation.

> The GrainTracker shouldn't be making that decision on it's own. I want it to be more general. This is why I like you creating a new class to do this work.

Your decision is wise, so I want to derive a new class called GrainTrackerMerge, which will be responsible for the merging of grains based on their crystallography.

Best Regards,
Wei
